CHANGES.md

```diff
@@ -22,6 +22,10 @@
 
   - <https://github.com/georust/gdal/pull/355>
 
+- Implemented `Geometry::make_valid` and `Geometry::make_valid_ex`.
```

Comment:
When reviewing the following changes, I'm interested in knowing what our policy should be here: 

a. Mirror `foo` & `foo_ex` patterns in C/C++ API
b. Have only `foo`, but always require `FooOpts`, providing an appropriate `FooOpts::default()` to minimize additional ceremony.

Question will make more sense in the context of what is to come.

Reply:
I'm inclined to prefer single functions instead of pairs like these. Note that GDAL already has some `Ex` functions, so us making up new ones might end up being quite confusing. But I guess that's not such a great risk, GDAL would have to add a new parameter that can't be expressed as an existing `CSLStringList`.

Reply:
What if the signature was something like `fn make_valid<O: Into<CslStringList>>(&self, opts: O)`? That way you could use `MakeValidOpts` if you want, or do it all yourself manually? 🤔 

(NB: I think figuring out the right pattern here is important, because looking ahead I see more of this sort of things coming up).

Reply:
Yeah, I thought about that, but it has some disadvantages:

 - it makes the function generic, increasing the compile times and bloating the code if we're not careful to split it in two
 - it's less discoverable, you're ready to use the function, then you realize you have to see what implements `Into<CslStringList>`. Examples can help here and I think rustdoc shows these implementations when you look at the trait, but it's still a bit distracting.

Reply:
@lnicola Would you mind taking a look at this latest version? [Here's how it looks in use](https://github.com/georust/gdal/pull/356/files#diff-4597c3ebc0daa4f8f578c4684bc8c616cb21d6ec4328936ff017c23444d7330cR691-R737).

TBH, because I don't have as much of an aversion to generics/monomorphism (when compared to the benefits), and think documentation can assist with discoverability, I much prefer the `Into<CslStringList>` approach to what is in this version. But I admittedly I have a bias toward ergonomics and have never had to deal with the consequences of code bloat. 

That said, I'll defer to you on what's ultimately best for the project; happy to follow your lead here. 😃 Note: If we *do* go with the generic `CslStringList` parameter version, I'll probably remove `MakeValidOpts` until we come up with a consistent approach to string-list-option construction.

Reply:
PS: I'll fix docs and move this out of draft once we settle on a design.

Reply:
PS: Just re-read this, and maybe should have drawn attention earlier to the fact that GDAL already has `OGR_G_MakeValid` and `OGR_G_MakeValidEx` variants. So our doubling with `Ex` functions in this case wouldn't technically be making up a new one.

> I'm inclined to prefer single functions instead of pairs like these. Note that GDAL already has some `Ex` functions, so us making up new ones might end up being quite confusing. But I guess that's not such a great risk, GDAL would have to add a new parameter that can't be expressed as an existing `CSLStringList`.

Reply:
@lnicola After sleeping on it, I'm now leaning toward this being the API, given the mirroring of `OGR_G_MakeValid`, and `OGR_G_MakeValidEx`.

```
pub fn make_valid(&self) -> Result<Geometry>
```
and
```
pub fn make_valid_ex(&self, opts: &CslStringList) -> Result<Geometry>
```

For now I'll drop `MakeValidOpts`, but possibly consider in the future a separate module of types for idiomatically constructing valid argument combinations and converting them to `CslStringList`, for methods that accept them.

WDYT?

Reply:
Sorry, I'll try to take another look tomorrow.

Reply:
> Yeah, but the C++ API only has

huh... you know, until now I just assumed the C API was the core with C++ built on top of that, and not the other way around! 🤦

Reply:
Nope, GDAL is C++, but it exposes a C API. I don't know if the C++ one is actually public.

Reply:
@jdroenner @rmanoka @ChristianBeilschmidt or anyone else: do you have opinions about the API approach we should take now an moving forward when optional arguments are specified with `CslStringList`? We're trying to decide a consistent approach as we expose more functions that use a list of `"KEY=VALUE"` strings to specify options.

Reply:
i think the `O: Into<X> ` pattern is really nice in many cases. However, in this case i think we can go with `Option<X>` or  `CslStringList::default()`. Since it is a list an option is somehow redundant. but maybe more ergonomic.

Reply:
@jdroenner does that trade-off change for you once we bring in `MakeValidOpts`? `make_valid(MakeValidOpts::Structure)` looks a bit nicer than `make_valid(MakeValidOpts::Structure.into())`.